subprojects/core/src/integTest/groovy/org/gradle/api/invocation/GradleLifecycleBeforeProjectEagerExecutionIntegrationTest.groovy

```diff
@@ -0,0 +1,245 @@
+            "task('foo')",
+            "defaultTasks",
+            "getAnt()",
+            "ant{}",
+            "createAntBuilder()",
+            "layout",
+            "repositories",
+            "dependencies",
+            "state",
+            "extensions",
+            "buildscript",
+            "configurations",
+            "components",
+            "artifacts",
+            "convention",
+            "plugins",
+            "pluginManager",
+            "getProperties()",
+            "findProperty('foo')",
+            "getProperties()",
+            "hasProperty('foo')",
+            "apply{}"
+        ]
+    }
+
+    def 'lifecycle.beforeProject is executed only once for a project'() {
+        given:
+        settingsFile """
+            rootProject.name = 'root'
+            gradle.lifecycle.beforeProject {
+                println "lifecycle.beforeProject for \${name}"
+            }
+            include(":a")
+        """
+
+        file("a/build.gradle") << ""
+
+        buildFile """
+            allprojects { $projectMutableStateAccess }
+            $secondEagerExecutionBlock
+        """
+
+        when:
+        run "help", "-q"
+
+        then:
+        output.count("lifecycle.beforeProject for root") == 1
+        output.count("lifecycle.beforeProject for a") == 1
+
+        where:
+        secondEagerExecutionBlock << [
+            "subprojects { $projectMutableStateAccess }",
+            "afterEvaluate { allprojects { $projectMutableStateAccess } }"
+        ]
+    }
+
+    def 'lifecycle.beforeProject can be executed before project.#api'() {
+        given:
+        settingsFile """
+            rootProject.name = 'root'
+            gradle.lifecycle.beforeProject {
+                ext.foo = "\$name bar"
+            }
+            include(":a")
+            include(":b")
+        """
+
+        file("a/build.gradle") << ""
+        file("b/build.gradle") << ""
+
+        buildFile"""
+            $invocation { project ->
+                println "\${project.foo}"
+            }
+        """
+
+        when:
+        run "help", "-q"
+
+        then:
+        outputContains expectedOutput
+
+        where:
+        api                | invocation                            | expectedOutput
+        "allprojects"      | "allprojects"                         | "root bar\na bar\nb bar"
+        "subprojects"      | "subprojects"                         | "a bar\nb bar"
+        "project"          | "project(':a')"                       | "a bar"
+        "findProject"      | "configure(findProject(':a'))"        | "a bar"
+        "getAllprojects"   | "getAllprojects().forEach"            | "root bar\na bar\nb bar"
+        "getSubprojects"   | "getSubprojects().forEach"            | "a bar\nb bar"
+        "getChildProjects" | "getChildProjects().values().forEach" | "a bar\nb bar"
+    }
+
+    def 'lifecycle.beforeProject can be executed before gradle.allprojects'() {
```

Comment:
There is a difference with the new callbacks between executing before and being triggered from inside of another callback by mutable state access.

Can you update this test to make the two cases more apparent?

Reply:
Added additional prints